resolver/resolver.cc

```diff
@@ -303,60 +303,120 @@ class ResolveConstantsWalk {
+                // already know this not a leaf.
+                res.exports.emplace_back(path[prefixLen]);
+            }
+            return res;
+        }
+
+        static PackageStub regularImport(const core::packages::PackageInfo &info) {
+            PackageStub res;
+            res.packageId = info.mangledName();
+            res.fullName = info.fullName();
+            return res;
+        }
+
         // Check that the candidate name is one of the top-level exported names from a parent package.
         bool exportsSymbol(core::NameRef candidate) const {
             return absl::c_find(this->exports, candidate) != this->exports.end();
         }
 
-        // Check that a symbol matches the full name of this parent package.
-        bool matchesOwners(const core::GlobalState &gs, core::ClassOrModuleRef sym) const {
-            for (auto it = this->fullName.rbegin(); it != this->fullName.rend(); ++it) {
-                if (!sym.exists()) {
-                    return false;
-                }
+        bool couldDefineChildNamespace(const core::GlobalState &gs, const std::vector<core::NameRef> &prefix,
+                                       const std::vector<ast::ConstantLit *> &suffix) const {
+            // This check is only valid for regular exports.
+            ENFORCE(this->exports.empty());
 
-                if (sym.data(gs)->name != *it) {
-                    return false;
+            // The reasoning is as follows: the prefix is derived from a nesting scope paired with the symbol being
+            // resolved. The nesting scopes could only define a package in the parent namespace, while this check is
+            // only applied to packages that are known to not occupy that part of the namespace.
+            if (this->fullName.size() <= prefix.size()) {
+                return false;
+            }
+
+            if (!std::equal(prefix.begin(), prefix.end(), this->fullName.begin())) {
+                return false;
+            }
+
+            auto it = this->fullName.begin() + prefix.size();
+            for (auto *cnst : suffix) {
+                if (it == this->fullName.end()) {
+                    return true;
                 }
 
-                auto owner = sym.data(gs)->owner;
-                if (!owner.isClassOrModule()) {
+                auto &original = ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(cnst->original);
+                if (original.cnst != *it) {
                     return false;
                 }
 
-                sym = owner.asClassOrModuleRef();
+                ++it;
             }
 
-            return sym == core::Symbols::root();
+            return true;
+        }
+
+        bool operator<(const PackageStub &other) const {
+            return absl::c_lexicographical_compare(this->fullName, other.fullName,
+                                                   [](auto l, auto r) { return l.rawId() < r.rawId(); });
         }
```

Comment:
Probably not? This is a hold over from when I thought we needed to find the best package for a symbol to stub, rather than mirror what the vm would do at runtime.